server/app/api/registry/registry_controller.py

```diff
@@ -25,17 +34,43 @@ def registry_endpoint(
 
     return registry_service.handle_registry_request(username, password, scopes, service)
 
-@router.api_route("/notifications", methods=["POST", "PUT"], summary="Webhook for Docker Registry")
+@internal_router.api_route("/notifications", methods=["POST", "PUT"], summary="Webhook for Docker Registry")
 def registry_notification_endpoint(data: dict, registry_service: RegistryService = Depends(get_registry_service)):
     for event in data["events"]:
         action = event.get("action", None)
         username = event.get("actor", {}).get("name", None)
         repository = event.get("target", {}).get("repository", None)
         tag = event.get("target", {}).get("tag", None)
-        
+        digest = event.get("target", {}).get("digest", None)
+        method = event.get("request", {}).get("method", None)
+        url = event.get("target", {}).get("url", None)
+
         try:
-            registry_service.on_notification(username, action, repository, tag)
+            registry_service.on_notification(username, action, repository, tag, digest, method, url)
         except Exception as e:
             LOGGER.error(f"Couldn't handle Distribution webhook: {e}")
 
-    return {}
\ No newline at end of file
+    return {}
+
+@external_router.delete("/tag", status_code=200, summary="Delete a tag by its name", response_model=DeleteTagResponseDTO)
+@pre_authorize([UserRole.user, UserRole.admin])                             
+async def delete_tag_endpoint(
+    jwt: JWTDep,
+    dto: DeleteTagDTO,
+    repo_service: RepositoryService = Depends(get_repo_service), 
+    registry_service: RegistryService = Depends(get_registry_service), 
+    tag_service: TagService = Depends(get_tag_service),
+    registry_client: RegistryClient = Depends(get_registry_client),
+    access_control_service: AccessControlService = Depends(get_access_control_service)
+):
+    user_id = get_id_from_jwt(jwt)
+    username = get_username_from_jwt(jwt)
+    repo = repo_service.find_by_id(dto.repo_id)
+    tag = tag_service.find_by_name_and_repo_id(dto.tag_name, dto.repo_id)
+
+    if not access_control_service.has_delete_tag_access(user_id, repo.id):
+        raise AccessDeniedException(f"User {user_id} cannot delete a tag {tag.name} of repository with identifier {repo.id}")
+
+    response = await registry_service.delete_tag(registry_client, username, repo, tag)   
+
```

Comment:
Why do we check for access control here instead of inside `registry_service.delete_tag()`?

Reply:
[Note: we came to the conclusion that it _should_ be in the service layer because it's business logic. `pre_authorize` being in the controller layer is a remnant of Spring Boot. Unfortunately, there are many other places where we broke this rule, and seeing as it's not a major issue, it's best to leave it alone and not create additional work]